server/src/main/scala/com/cloudera/livy/server/LivyServer.scala

```diff
@@ -67,8 +69,17 @@ class LivyServer extends Logging {
     // Initialize YarnClient ASAP to save time.
     if (livyConf.isRunningOnYarn()) {
       Future { SparkYarnApp.yarnClient }
+    } else {
+      // If recovery is turned on but we are not running on YARN, quit.
+      require(livyConf.get(LivyConf.RECOVERY_MODE) == SessionRecoveryMode.OFF,
```

Comment:
From our point of view, it's a critical issue because users expect session to persist across reboot when they turn on recovery. If Livy can't do it, it should return a fatal error so it doesn't surprise our users.